File: tests/rank/test_voting_methods.py
from __future__ import annotations

import numpy as np
import pytest

from scorio import rank


@pytest.mark.parametrize(
    ("fn", "kwargs"),
    [
        (rank.borda, {}),
        (rank.copeland, {}),
        (rank.win_rate, {}),
        (rank.minimax, {"variant": "margin", "tie_policy": "half"}),
        (rank.schulze, {"tie_policy": "half"}),
        (rank.ranked_pairs, {"strength": "margin", "tie_policy": "half"}),
        (rank.kemeny_young, {"tie_policy": "half", "time_limit": 1.0}),
        (rank.nanson, {"rank_ties": "average"}),
        (rank.baldwin, {"rank_ties": "average"}),
        (rank.majority_judgment, {}),
    ],
)
def test_voting_methods_smoke_and_ordering(
    ordered_binary_R: np.ndarray,
    rank_assertions,
    fn,
    kwargs: dict,
) -> None:
    ranking, _ = rank_assertions.assert_ranking_and_scores(
        fn(ordered_binary_R, return_scores=True, **kwargs)
    )
    rank_assertions.assert_ordering_sanity(ranking, best_idx=0, worst_idx=3)


def test_voting_option_branches(
    ordered_binary_small_R: np.ndarray, rank_assertions
) -> None:
    out_minimax = rank.minimax(
        ordered_binary_small_R,
        variant="winning_votes",
        tie_policy="ignore",
        return_scores=True,
    )
    out_ranked_pairs = rank.ranked_pairs(
        ordered_binary_small_R,
        strength="winning_votes",
        tie_policy="ignore",
        return_scores=True,
    )
    out_kemeny = rank.kemeny_young(
        ordered_binary_small_R,
        tie_policy="ignore",
        tie_aware=False,
        time_limit=1.0,
        return_scores=True,
    )

    rank_assertions.assert_ranking_and_scores(out_minimax)
    rank_assertions.assert_ranking_and_scores(out_ranked_pairs)
    rank_assertions.assert_ranking_and_scores(out_kemeny)


def test_voting_validation_errors(ordered_binary_small_R: np.ndarray) -> None:
    with pytest.raises(ValueError, match="variant must be one of"):
        rank.minimax(ordered_binary_small_R, variant="bad")

    with pytest.raises(ValueError, match="strength must be one of"):
        rank.ranked_pairs(ordered_binary_small_R, strength="bad")

    with pytest.raises(ValueError, match="tie_policy must be one of"):
        rank.schulze(ordered_binary_small_R, tie_policy="bad")

    with pytest.raises(ValueError, match="time_limit must be a positive finite scalar"):
        rank.kemeny_young(ordered_binary_small_R, time_limit=0.0)

    with pytest.raises(ValueError, match='unknown method "bad"'):
        rank.nanson(ordered_binary_small_R, rank_ties="bad")

File: julia/Scorio.jl/test/rank/test_voting_methods.jl
using Test
using Scorio

@testset "rank/test_voting_methods.jl" begin
    R = ordered_binary_R()
    R_small = ordered_binary_small_R()

    @testset "voting methods smoke and ordering" begin
        calls = [
            () -> Scorio.Rank.borda(R; return_scores=true),
            () -> Scorio.Rank.copeland(R; return_scores=true),
            () -> Scorio.Rank.win_rate(R; return_scores=true),
            () -> Scorio.Rank.minimax(R; variant="margin", tie_policy="half", return_scores=true),
            () -> Scorio.Rank.schulze(R; tie_policy="half", return_scores=true),
            () -> Scorio.Rank.ranked_pairs(
                R;
                strength="margin",
                tie_policy="half",
                return_scores=true,
            ),
            () -> Scorio.Rank.kemeny_young(
                R;
                tie_policy="half",
                time_limit=1.0,
                return_scores=true,
            ),
            () -> Scorio.Rank.nanson(R; rank_ties="average", return_scores=true),
            () -> Scorio.Rank.baldwin(R; rank_ties="average", return_scores=true),
            () -> Scorio.Rank.majority_judgment(R; return_scores=true),
        ]

        for run in calls
            ranking, _ = assert_ranking_and_scores(run())
            assert_ordering_sanity(ranking; best_idx=1, worst_idx=4)
        end
    end

    @testset "option branches" begin
        out_minimax = Scorio.Rank.minimax(
            R_small;
            variant="winning_votes",
            tie_policy="ignore",
            return_scores=true,
        )
        out_ranked_pairs = Scorio.Rank.ranked_pairs(
            R_small;
            strength="winning_votes",
            tie_policy="ignore",
            return_scores=true,
        )
        out_kemeny = Scorio.Rank.kemeny_young(
            R_small;
            tie_policy="ignore",
            tie_aware=false,
            time_limit=1.0,
            return_scores=true,
        )

        assert_ranking_and_scores(out_minimax)
        assert_ranking_and_scores(out_ranked_pairs)
        assert_ranking_and_scores(out_kemeny)
    end

    @testset "validation errors" begin
        @test_throws ErrorException Scorio.Rank.minimax(R_small; variant="bad")
        @test_throws ErrorException Scorio.Rank.ranked_pairs(R_small; strength="bad")
        @test_throws ErrorException Scorio.Rank.schulze(R_small; tie_policy="bad")
        @test_throws ErrorException Scorio.Rank.kemeny_young(R_small; time_limit=0.0)
        @test_throws ErrorException Scorio.Rank.nanson(R_small; rank_ties="bad")
    end
end
